### DISCUSSION TITLE
Booting iPXE over WAN

### DISCUSSION THREAD

**Discussion by Theknight2015** [ipxe/ipxe#835] [Q&A]
Hello everyone. I'm new to iPXE and don't have a whole lot of experience with PXE booting in general but I have been experimenting with iPXE with some pretty good success.

I will go ahead and jump right in to the details of what I have going on and what I'm having trouble understanding/making work.

I have setup a thin client server using a program called OpenThinClient(OTC) 2022 on a windows 2016 Server. The installation allows you to set custom configs of your OTC server using port 8080 to login to a web portal and adjust the settings from there. It also allows you control over the pxe clients in that you have to pxe boot a potential client, the MAC is seen on the network and by the OTC server and you add/configure the device using its MAC in the portal.

I wanted to use iPXE to boot the thin clients to that server so I could potentially boot it from outside my home network as well. I want to do this because the thin client is setup with pre-defined user applications and settings that no matter what PC I use, if my account logs in, it will pickup those user settings as though it was the same PC. 

My hope was to take a small intel compute stick with me on trips and while I travel for work so I can plug the compute stick in to a hotel TV and it boot to the thin client server. This would allow me to have the same settings of the thin client and all the pre-defined applications load no matter where I was. Then I was hoping to be able to click on a pre-defined RDP session and it open to the corresponding host device that has my streaming accounts on it.

The basic operation would be -
- Intel Compute Stick (Remote Thin Client) turns on and uses a .iso burned on a USB to boot iPXE
- The embedded script in the iPXE .iso USB tells the client to pull the files from my home OTC server for booting over tftp or whatever method would work best for my case (Currently trying with tftp)
- The Remote Thin Client would pull those files from the OTC Server and use them to boot
- The system would boot to the thin client environment where it would load the settings from my OTC Server
- I would then sign in and click on my RDP session to mt Home Desktop that would then open and allow me to use it remotely

I have been able to setup a Debian 11 Instance and get iPXE installed and I have been able to create a .iso file using the make command for creating an .iso
I then coppied the .iso to a windows machine and used Rufus to burn it onto a USB drive that I plugged into a PC on the network.

**Local Network Boot:**
Using this script, I was able to boot the OTC server and gain access to my settings and RDP session on my home network. 
The DHCP server is on:
192.168.1.253
The root directory("/openthinclient") of the OTC server is on:
192.168.1.145
The DHCP server has option 66 set to "192.168.1.145" and option 67 set to "/openthinclient"

Embedded Script In iPXE.iso Build File:
---------------------------------------------------------------------------------------------------------------------------------------------------
#!ipxe
dhcp net0
set next-server 192.168.1.253
set 150:string ${next-server}
set 67:string 01${mac:hex}
kernel tftp://192.168.1.145/lpxelinux.0

boot
---------------------------------------------------------------------------------------------------------------------------------------------------

**This process works great and the system boots as it normally would when not using iPXE.** 

**My problem arises when I try the same thing on an "External Network" (A second ISP provided dedicated line to my house with a different public IP and it's own Router)** - No I'm not trying to boot to the OTC server using the second line, I'm trying to boot the Remote Thin Client using it. I'm doing this so the Remote Thin Client is technically not on my home network but rather acting as though it's on a remote network like a hotel or restaurant or something.

I use the following script where XXX.XXX.XXX.XXX is the IP of my home network's public IP and the router is forwarding all tftp port 69 traffic to my 192.168.1.145 OTC server.
---------------------------------------------------------------------------------------------------------------------------------------------------
#!ipxe
dhcp net0
set next-server XXX.XXX.XXX.XXX
set 150:string ${next-server}
set 67:string 01${mac:hex}
kernel tftp://XXX.XXX.XXX.XXX:69/lpxelinux.0

boot
---------------------------------------------------------------------------------------------------------------------------------------------------

**I would think this solution works as it successfully grabs the file but after that it just hangs and does nothing. After about 10 minutes I get an error saying "Failed to load ldlinux.c32" and there is nothing after that.

I looked into it a bit and apparently the OTC Server files are only transferable through tftp according to OTC's documentation. I would assume this isn't a problem since I'm using tftp to get the file and they are retrieved successfully.

I think I have a major embedded script issue seeing as I do not know how to format a lot of it or what it all does.

Any help or advice on this would be appreciated.**

**Another solution I thought of was to use iPXE to get a lease from the DHCP home server if that is possible and run as though the Remote Thin client is on the local home network when it's not actually there but I'm not sure how I would do that.**

**My last solution was to use iPXE and some embedded script to create a VPN tunnel from the Remote thin Client to my home network if that was possible?**

**Comment by NiKiZe** [comment]
I would recommend building ipxe.usb instead of the .iso, since that is what you want.

Tftp works on LAN, but going over any border and it goes flakey, use http instead.

You are chaining to syslinux, I would do the full load in iPXE script.

DHCP etc on your home network is irrelevant (see it as unreachable)
You should also assume that NFS is unreachable, and make sure you load everything over http(s).

**Reply by Theknight2015** [reply]
NiKiZe - I took your advice and created a UEFI bootable USB. I then played around with the settings of what file I was using in order to boot the system. I found that in a UEFI environment, the system won't boot with the original lpxelinux.0 file but rather boots using the ipxe64.efi file which makes sense as it's now operating in a UEFI environment. On local LAN connection, it works flawlessly but unfortunately on WAN I'm now getting a new error saying
--------------------------------------------------------------------------------------------------------------------------------------------------------

DHCP Connecting...... OK
tftp://XXX.XXX.XXX.XXX/ipxe64.efi....OK
Option 66 Not Set
No response from openthinclient server. Please contact your administrator.

--------------------------------------------------------------------------------------------------------------------------------------------------------

Now my question is how do I go around the option 66 requirement when I don't even have it plugged into the script? I'm guessing what I really need to do is figure out a way to connect to the DHCP server on my local network???? That way it picks up the option 66 parameter? Or is there a way to hardcode iPXE option 66 to a http address?

The reason I'm still using tftp on the initial loading of ipxe64.efi is that when I switch it to http, it doesn't work. It just says connection timed out and resets the PC to boot again.

**Reply by NiKiZe** [reply]
DHCP is a local protocol sending raw packets from/to non defined IPs.
You do not event want lpxelinux. Just iPXE, and again use http instead of tftp.

If you have issues with http, then you need to debug that, tcpdump or wireshark is great tools.

You can also boot into the iPXE shell. Run dhcp to get network and then chain http://boot.ipxe.org/demo
Or chain http://gentoo.ipxe.se/boot.ipxe

This is 2 different live Linux systems that should prove that you can boot over http. From here you can go back and get your own http server running.

I don't know where or how you got to your current setup, but the sooner you can avoid tftp the better. Use the features of iPXE instead of trying to chain into some other PXE bootloader that just complicates things for you.

**Comment by stappersg** [comment]
> I'm new to iPXE and don't have a whole lot of experience with PXE booting in general but I have been experimenting with iPXE with some pretty good success.

I do read that in the context of this discussion as:
> Pretty good success with iPXE on a LAN, but new to networking when there are severals LANs involved.

 (And several involved LANs it is named  "WAN" ;-)


@Theknight2015 verify the various WAN connections without _iPXE_. And when you have those Wide Area Network connections working (debugged and fixed) then put _iPXE_ back in the loop. You while learn a lot during your journey, do [ask smart questions](http://www.catb.org/~esr/faqs/smart-questions.html) to get the most out of it.
